File: java/bank-account/src/main/java/BankAccount.java
class BankAccount {

    private int balance = 0;
    private boolean isOpen = false;

    void open() throws BankAccountActionInvalidException {
        if (isOpen) {
            throw new BankAccountActionInvalidException("Account already open");
        } else {
            isOpen = true;
            balance = 0;
        }
    }

    void close() throws BankAccountActionInvalidException {
        if (isOpen) {
            isOpen = false;
            balance = 0;
        } else {
            throw new BankAccountActionInvalidException("Account not open");
        }
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        verifyOpen();
        return balance;

    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        verifyOpen();
        if (verifyTxAmount(amount)) {
            balance += amount;
        }
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        verifyOpen();
        if ((balance - amount) < 0) {
            throw new BankAccountActionInvalidException("Cannot withdraw more money than is currently in the account");
        } else {
            verifyTxAmount(amount);
            balance -= amount;
        }

    }

    synchronized void verifyOpen() throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account closed");
        }
    }

    synchronized boolean verifyTxAmount(int amount) throws BankAccountActionInvalidException {
        if (amount > 0) {
            return true;
        } else {
            throw new BankAccountActionInvalidException("Cannot deposit or withdraw negative amount");
        }
    }

}

File: go/bank-account/bank_account.go
package account

import "sync"

// Account represents a bank account with a balance,
// its status of being open, and a mutex
type Account struct {
	balance int64
	open    bool
	*sync.Mutex
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	return &Account{amount, true, &sync.Mutex{}}
}

func (a *Account) Balance() (int64, bool) {
	a.Lock()
	defer a.Unlock()
	return a.balance, a.open
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.Lock()
	defer a.Unlock()
	if !a.open {
		return 0, false
	}
	if amount+a.balance < 0 {
		return a.balance, false
	}
	a.balance += amount
	return a.balance, true
}

func (a *Account) Close() (int64, bool) {
	a.Lock()
	defer a.Unlock()
	if !a.open {
		return a.balance, false
	}
	bal := a.balance
	a.balance = 0
	a.open = false
	return bal, true
}
